platforms/core-configuration/model-core/src/main/java/org/gradle/internal/properties/bean/DefaultPropertyWalker.java

```diff
@@ -127,7 +158,13 @@ public TaskDependencyContainer getTaskDependencies() {
         public void maybeFinalizeValue() {
             if (isConfigurable()) {
                 Object value = cachedInvoker.get();
-                ((HasConfigurableValueInternal) value).implicitFinalizeValue();
+                if (isUpgradedProperty) {
+                    // Upgraded properties should not be finalized to simplify migration.
+                    // This behaviour should be removed with Gradle 10.
+                    ((HasConfigurableValueInternal) value).warnOnUpgradedPropertyChanges(propertyDisplayName);
```

Comment:
💭 I left "Upgraded" context in method names so we know this is used only property upgrades. I think we don't need a general solution